modules/caddyhttp/celmatcher.go

```diff
@@ -83,14 +84,39 @@ func (m *MatchExpression) UnmarshalJSON(data []byte) error {
 }
 
 // Provision sets ups m.
-func (m *MatchExpression) Provision(_ caddy.Context) error {
+func (m *MatchExpression) Provision(ctx caddy.Context) error {
 	// replace placeholders with a function call - this is just some
 	// light (and possibly naïve) syntactic sugar
 	m.expandedExpr = placeholderRegexp.ReplaceAllString(m.Expr, placeholderExpansion)
 
 	// our type adapter expands CEL's standard type support
 	m.ta = celTypeAdapter{}
 
+	// initialize the CEL libraries from the Matcher implementations which
+	// have been configured to support CEL.
+	matcherLibProducers := []celLibraryProducer{
+		MatchHeader{},
+		MatchHost{},
+		MatchMethod{},
+		MatchPath{},
+		MatchPathRE{},
+		MatchProtocol(""),
+		MatchQuery{},
+	}
```

Comment:
Right, so instead of hardcoding this here, I think what we should do is call `caddy.GetModules("http.matchers")` which will return a list of all the matcher modules, then they can be type asserted to check if they are a `celLibraryProducer`. That way, it doesn't need to be hard-coded, and any third-party plugin could add support. Obviously this would mean we'd need to export `CelLibraryProducer` so it can be used by third-party modules.

Reply:
I was thinking of leaving third-party support out until all of the standard matchers had been converted and you'd gotten some feature requests from plugin writers to open up the functionality. Basically, I didn't want to commit Caddy to supporting an expanded API surface for matchers until you're convinced it's ready, safe, and desirable for others to use. Thoughts?

I'll happily change the call to use `caddy.GetModules` though.

Reply:
I think we can just put in the godoc comment that it's still `EXPERIMENTAL` as a warning to potential users. It's what we've done for other features that aren't yet final.

Reply:
The CEL related features are still marked experimental. Do you want a separate callout for the `CELLibraryProducer`? Otherwise all comments have been addressed and I've gotten the code working with placeholders as well.

Reply:
IMO we can probably remove the experimental marker on CEL/`expression` at this point, I don't think it's going away, and it's seemed pretty stable thus far. WDYT @mholt?

But yeah we can mark `CELLibraryProducer` itself experimental if we do. Either way, there aren't many third-party request matchers yet so I doubt it'll matter much.